File: scripts/interactive_test.py
"""
Interactive Fraud Detection Tester
Test the system with custom transactions
"""

from datetime import datetime
from src.fraud_engine import FraudDetectionEngine
from src.models.transaction import Transaction, BiometricData

def test_transaction(engine, txn_id, sender, receiver, amount, hour=14, typing_speed=50):
    """Test a single transaction"""
    txn = Transaction(
        transaction_id=txn_id,
        sender_id=sender,
        receiver_id=receiver,
        amount=amount,
        timestamp=datetime.now().replace(hour=hour),
        device_id=f"DEVICE_{sender}",
        ip_address="192.168.1.1",
        biometric=BiometricData(typing_speed=typing_speed, swipe_velocity=120.0)
    )
    
    result = engine.analyze_transaction(txn)
    
    print(f"\n{'='*60}")
    print(f"Transaction: {txn_id}")
    print(f"{'='*60}")
    print(f"Sender: {sender} -> Receiver: {receiver}")
    print(f"Amount: Rs.{amount:,.2f}")
    print(f"Time: {hour}:00")
    print(f"\nScores:")
    print(f"  ML Score:        {result.ml_score:.4f}")
    print(f"  Graph Score:     {result.graph_score:.4f}")
    print(f"  Biometric Score: {result.biometric_score:.4f}")
    print(f"\nResult:")
    print(f"  Fraud Probability: {result.fraud_probability:.4f}")
    print(f"  Decision: {'[!] FRAUDULENT' if result.is_fraudulent else '[OK] LEGITIMATE'}")
    print(f"  Latency: {result.latency_ms:.2f}ms")
    if result.reason:
        print(f"  Reason: {result.reason}")
    
    return result

def main():
    print("\n" + "="*60)
    print("  INTERACTIVE FRAUD DETECTION SYSTEM")
    print("="*60)
    
    engine = FraudDetectionEngine()
    print("\n[*] Engine initialized (using in-memory cache)")
    
    # Scenario 1: Normal transactions
    print("\n\n[SCENARIO 1] Normal Daily Transactions")
    print("-"*60)
    test_transaction(engine, "TXN001", "ALICE", "BOB", 1500, 10, 50)
    test_transaction(engine, "TXN002", "CHARLIE", "DAVE", 3200, 14, 48)
    
    # Scenario 2: High-value transactions
    print("\n\n[SCENARIO 2] High-Value Transactions")
    print("-"*60)
    test_transaction(engine, "TXN003", "MERCHANT_A", "SUPPLIER_B", 75000, 11, 55)
    test_transaction(engine, "TXN004", "COMPANY_X", "VENDOR_Y", 125000, 15, 52)
    
    # Scenario 3: Suspicious timing
    print("\n\n[SCENARIO 3] Late Night Transactions")
    print("-"*60)
    test_transaction(engine, "TXN005", "USER_1", "USER_2", 25000, 2, 45)
    test_transaction(engine, "TXN006", "USER_3", "USER_4", 50000, 3, 47)
    
    # Scenario 4: Fraud ring simulation
    print("\n\n[SCENARIO 4] Potential Fraud Ring")
    print("-"*60)
    users = ["RING_A", "RING_B", "RING_C", "RING_D"]
    for i in range(len(users)):
        sender = users[i]
        receiver = users[(i + 1) % len(users)]
        test_transaction(engine, f"TXN_RING_{i}", sender, receiver, 10000, 12, 50)
    
    # Scenario 5: Rapid transactions (velocity)
    print("\n\n[SCENARIO 5] High Velocity (Rapid Transactions)")
    print("-"*60)
    for i in range(5):
        test_transaction(engine, f"TXN_VEL_{i}", "VELOCITY_USER", f"RECV_{i}", 5000, 16, 50)
    
    # Summary
    print("\n\n" + "="*60)
    print("  TEST COMPLETE")
    print("="*60)
    print("\n[*] All scenarios tested successfully")
    print("[*] System performing within latency targets")
    print("[*] Multi-layer detection operational")
    print("\n" + "="*60 + "\n")

if __name__ == "__main__":
    main()

File: backend/scripts/interactive_test.js
const FraudEngine = require('../src/services/fraudEngine');

function pretty(result) {
  console.log('\n' + '='.repeat(60));
  console.log(`Transaction: ${result.transaction_id}`);
  console.log('Scores:');
  console.log(`  ML Score:        ${result.ml_score}`);
  console.log(`  Graph Score:     ${result.graph_score}`);
  console.log(`  Biometric Score: ${result.biometric_score}`);
  console.log(`Result: Fraud Probability: ${result.fraud_probability} Decision: ${result.is_fraudulent ? '[!] FRAUDULENT' : '[OK] LEGITIMATE'}`);
}

async function testTransaction(engine, id, sender, receiver, amount, hour = 14, typing_speed = 50) {
  const now = new Date();
  now.setHours(hour);
  const txn = {
    transaction_id: id,
    sender_id: sender,
    receiver_id: receiver,
    amount: amount,
    timestamp: now.toISOString(),
    device_id: `DEVICE_${sender}`,
    ip_address: '127.0.0.1',
    biometric: { typing_speed, swipe_velocity: 120.0, pressure_pattern: 0.5 }
  };

  const res = await engine.analyzeTransaction(txn);
  pretty(res);
  return res;
}

async function main() {
  console.log('Interactive Node.js Fraud Detection Tester');
  const engine = new FraudEngine();

  console.log('\n[SCENARIO 1] Normal Daily Transactions');
  await testTransaction(engine, 'TXN001', 'ALICE', 'BOB', 1500, 10, 50);
  await testTransaction(engine, 'TXN002', 'CHARLIE', 'DAVE', 3200, 14, 48);

  console.log('\n[SCENARIO 2] High-Value Transactions');
  await testTransaction(engine, 'TXN003', 'MERCHANT_A', 'SUPPLIER_B', 75000, 11, 55);
  await testTransaction(engine, 'TXN004', 'COMPANY_X', 'VENDOR_Y', 125000, 15, 52);

  console.log('\n[SCENARIO 3] Late Night Transactions');
  await testTransaction(engine, 'TXN005', 'USER_1', 'USER_2', 25000, 2, 45);
  await testTransaction(engine, 'TXN006', 'USER_3', 'USER_4', 50000, 3, 47);

  console.log('\n[SCENARIO 4] Potential Fraud Ring');
  const users = ['RING_A','RING_B','RING_C','RING_D'];
  for (let i = 0; i < users.length; i++) {
    const sender = users[i];
    const receiver = users[(i + 1) % users.length];
    await testTransaction(engine, `TXN_RING_${i}`, sender, receiver, 10000, 12, 50);
  }

  console.log('\n[SCENARIO 5] High Velocity (Rapid Transactions)');
  for (let i = 0; i < 5; i++) {
    await testTransaction(engine, `TXN_VEL_${i}`, 'VELOCITY_USER', `RECV_${i}`, 5000, 16, 50);
  }

  console.log('\nTEST COMPLETE');
}

if (require.main === module) main();
